### PR TITLE
Rewrite CosmosVFS

### PR DESCRIPTION
**Pull request by MishaTY**
Rewrites CosmosVFS class with a new API. You can easily get all of the disks and partitions inside of those disks.

TODO:

-  [x] Add better partition support
- [x] Add partition to disk
- [x] Remove partition
- [x] Add remove/add partitions functions to Disk.cs

### CODE REVIEW

File: source/Cosmos.HAL2/BlockDevice/Partition.cs
Comment (quajak) [inline source/Cosmos.HAL2/BlockDevice/Partition.cs:13]: Please rename this to `Host`

File: source/Cosmos.System2/FileSystem/CosmosVFS.cs
Comment (quajak) [inline source/Cosmos.System2/FileSystem/CosmosVFS.cs:44]: This should be suppressible, in which case an exception is thrown for the user os to handle.

File: source/Cosmos.System2/FileSystem/CosmosVFS.cs
Comment (quajak) [inline source/Cosmos.System2/FileSystem/CosmosVFS.cs:740]: Can you please refractor the getting and checking of the filesystem into a helper function to reduce repeated code?

File: source/Cosmos.System2/FileSystem/CosmosVFS.cs
Comment (quajak) [inline source/Cosmos.System2/FileSystem/CosmosVFS.cs:757]: Do we need these wrapper functions? Or it enough if you can get the file system from a drive id to query the data yourself?

File: source/Cosmos.System2/FileSystem/Disk.cs
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:39]: Please rename to `RegisteredFileSystemsTypes`

File: source/Cosmos.System2/FileSystem/ManagedPartition.cs
Comment (quajak) [inline source/Cosmos.System2/FileSystem/ManagedPartition.cs:45]: Maybe also add a version of the Mount function where you pass the FilesystemFactory in as well

**Review by quajak — CHANGES_REQUESTED** [review]
Thanks for the PR! The idea is solid, just a bit of tweaking/polishing left. Could you please also add some tests to the filesystem testrunner?

File: source/Cosmos.System2/FileSystem/Disk.cs
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/Disk.cs:17]: Oops, I forgot to make that private.

File: source/Cosmos.System2/FileSystem/Disk.cs
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/Disk.cs:29]: No, we don't. I was just copying and pasting code from the old CosmosVFS and forgot to remove that.

File: source/Cosmos.System2/FileSystem/Disk.cs
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/Disk.cs:24]: Ok, I will move that.

File: source/Cosmos.System2/FileSystem/CosmosVFS.cs
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/CosmosVFS.cs:751]: I think it's better to completely remove the VFSBase and CosmosVFS classes and replace them with a DiskManager class where the user can get all of the disks.

**Comment by MishaTY** [comment]
This PR is ready for review and merging now.

File: source/Cosmos.HAL2/BlockDevice/IDE.cs
Comment (quajak) [inline source/Cosmos.HAL2/BlockDevice/IDE.cs:56]: Can this be a for loop?

File: source/Cosmos.HAL2/BlockDevice/IDE.cs
Comment (quajak) [inline source/Cosmos.HAL2/BlockDevice/IDE.cs:58]: Also does the current `GetPartitions` allow for null partitions to happen?

File: source/Cosmos.HAL2/BlockDevice/IDE.cs
Comment (quajak) [inline source/Cosmos.HAL2/BlockDevice/IDE.cs:90]: Same question here

**Review by quajak — APPROVED** [review]
Looks good. I think charles should take a look as well before we merge.

**Comment by charlesbetros** [comment]
I think we should keep the VFS base class and use it everywhere. This way users can implement their own VFS if they want.

**Comment by MishaTY** [comment]
I didn't remove the VFSBase class. Users can also implement custom drive letters now.

File: source/Cosmos.System2/FileSystem/Disk.cs
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:36]: Why is this required? Isnt accessing the `PartitioningType` enough?

File: source/Cosmos.System2/FileSystem/Disk.cs
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:24]: Is the list necessary here? Cant we access the information from PartitioningType already? Also if we create a new Partition via the PartitioningType how do we update the list here?

File: source/Cosmos.System2/FileSystem/Disk.cs
Comment (quajak) [inline source/Cosmos.System2/FileSystem/Disk.cs:10]: Can we merge `PartitioningType` and `Disk` into the same class, since they seem to overlap a lot.

**Review by quajak — CHANGES_REQUESTED** [review]

File: source/Cosmos.System2/FileSystem/Disk.cs
Comment (MishaTY) [reply source/Cosmos.System2/FileSystem/Disk.cs:10]: We can, but the code will be messy if we do that.

Comment (quajak) [reply source/Cosmos.System2/FileSystem/Disk.cs:10]: How so?

**Pull request closed without merging**
